File: game-engine/NumberValidator.java
public class NumberValidator {

    public boolean isValidNumber(String number) {

        // Check null
        if (number == null) {
            return false;
        }

        // Check length
        if (number.length() != 4) {
            return false;
        }

        // Check digits only
        for (char ch : number.toCharArray()) {
            if (!Character.isDigit(ch)) {
                return false;
            }
        }

        // Check first digit is not 0
        if (number.charAt(0) == '0') {
            return false;
        }

        // Check unique digits
        for (int i = 0; i < number.length(); i++) {
            for (int j = i + 1; j < number.length(); j++) {
                if (number.charAt(i) == number.charAt(j)) {
                    return false;
                }
            }
        }

        return true;
    }
}

File: src/services/NumberValidator.ts
export class NumberValidator {

    static isValid(number: string): boolean {

        // Must be exactly 4 digits
        if (number.length !== 4) {
            return false;
        }

        // Must contain only digits
        if (!/^\d+$/.test(number)) {
            return false;
        }

        // First digit cannot be 0
        if (number[0] === "0") {
            return false;
        }

        // All digits must be unique
        return new Set(number).size === 4;
    }
}